src/main/java/com/aws/greengrass/componentmanager/plugins/docker/DockerImageDownloader.java

```diff
@@ -188,4 +140,113 @@ public boolean canUnarchiveArtifact() {
         // Not applicable for docker images since docker engine abstracts this
         return false;
     }
+
+    private void checkDownloadPrerequisites() throws PackageDownloadException {
+        // Check that Docker engine is installed
+        if (!dockerClient.dockerInstalled()) {
+            throw new PackageDownloadException(getErrorString(DOCKER_NOT_INSTALLED_ERROR_MESSAGE));
+        }
+    }
+
+    private File performDownloadSteps(Image image) throws PackageDownloadException, InterruptedException {
+        AtomicBoolean credentialRefreshNeeded = new AtomicBoolean(false);
+        do {
+            // Attempt getting credentials and login only for private registries, .
+            if (image.getRegistry().isPrivateRegistry()) {
+                // Get credentials for the registry
+                try {
+                    RetryUtils.runWithRetry(infiniteAttemptsRetryConfig, () -> {
+                        // Currently we only support private registries in ECR so assume others are public,
+                        // when we add support for private non-ECR registries, this can be expanded to retrieve
+                        // credentials for them on case by case basis
+                        if (image.getRegistry().isEcrRegistry()) {
+                            // Get auth token for ECR, which represents ECR registry credentials
+                            Registry.Credentials credentials =
+                                    ecrAccessor.getCredentials(image.getRegistry().getRegistryId());
+                            image.getRegistry().setCredentials(credentials);
+                            credentialRefreshNeeded.set(false);
+                        }
+                        return null;
+                    }, "get-ecr-auth-token", logger);
+                } catch (InterruptedException e) {
+                    throw e;
+                } catch (Exception e) {
+                    throw new PackageDownloadException(getErrorString("Failed to get auth token for docker login"), e);
+                }
+
+                // Login to registry
+                // TODO: [P44950158]: Avoid logging into registries which might already have been logged in previously
+                //  with the same and valid credentials by maintaining a cache across artifacts and deployments
+                run(() -> {
+                    if (credentialsUsable(image)) {
+                        return dockerClient.login(image.getRegistry());
+                    }
+                    // Credentials have expired, re-fetch and login again
+                    logger.atInfo().kv("registry-endpoint", image.getRegistry().getEndpoint())
+                            .log("Registry credentials have expired, fetching fresh credentials and logging in again");
+                    credentialRefreshNeeded.set(true);
+                    return false;
```

Comment:
does the return value get used anywhere? I don't think so right? It could be `return null` in both places?
I just checked and login can never return false, it will only throw an exception. Can you make this more obvious? either return null or add some comment here that the return value is not used.

It is confusing to read